### PR TITLE
feat(backend): configs for additional instance of mariadb and backend on same server

### PR DESCRIPTION
**Pull request by clauspeterhuebner**
### Issues

- None

### Todo

This is a description of actions done within this PR on local machine outside the git-control:
- [X] as prerequisit from the directory "gradido" 
   - [X] a copy of the subdirectories "database", "mariadb" and "backend/ are created in the new directory "gradido2"
   - [X] a copy of the files ".env.local.dist", ".env.shell", "docker-compose.override.yml", "docker-compose.test.yml", "docker-compose.yml", "package.json" and "yarn.lock" are created in the directory "gradido2"
- [X] All changes of this PR are made in the new directory "gradido2" and are outside of the normal git-control to avoid any influence on the default configurations
- [X] the changes are then shifted under git-control in the directory "gradido/separate_instance" to keep them outside of the default configurations

### CODE REVIEW

**Review by ulfgebhardt — COMMENTED** [review]
I do not understand the purpose of this PR - if we need a second instance I clone the repository again at a different place and adjust the configs or am I wrong here?

**Comment by clauspeterhuebner** [comment]
But where do you conserve the necessary changes of configuration for this additional instance on the same machine but different place?
If all the changes are done on the original files under git-control, you will overwrite the standard configuration as soon as you merge them.
You could create a separate document manually or create a script to change the necessary setting automatically.
But at the moment this was the KISS solution to save and communicate the changes for an additional instance running on the same machine.

File: separate_instance/backend/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,109 @@
+##################################################################################
```
Comment (ulfgebhardt) [inline separate_instance/backend/Dockerfile:1]: what is the change in this file?

File: separate_instance/database/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,128 @@
+##################################################################################
```
Comment (ulfgebhardt) [inline separate_instance/database/Dockerfile:1]: whats the change in this file?

File: separate_instance/database/src/config/index.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// ATTENTION: DO NOT PUT ANY SECRETS IN HERE (or the .env)
```
Comment (ulfgebhardt) [inline separate_instance/database/src/config/index.ts:1]: whats the change in this file?

File: separate_instance/docker-compose.override.yml [hunk 4]
```diff
@@ -0,0 +1,130 @@
+version: "3.4"
+
+services:
+
+#  ########################################################
+#  # FRONTEND #############################################
+#  ########################################################
+#  frontend:
```
Comment (ulfgebhardt) [inline separate_instance/docker-compose.override.yml:8]: no need to comment this out you can start stuff specifically `docker-compose up mariadb database backend`

File: separate_instance/docker-compose.override.yml [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#      # This makes sure the docker container has its own node modules.
+#      # Therefore it is possible to have a different node version on the host machine
+#      - frontend_node_modules:/app/node_modules
+#      # bind the local folder to the docker to allow live reload
+#      - ./frontend:/app
+#
+#  ########################################################
+#  # ADMIN INTERFACE ######################################
+#  ########################################################
+#  admin:
```
Comment (ulfgebhardt) [inline separate_instance/docker-compose.override.yml:26]: no need to comment this out you can start stuff specifically `docker-compose up mariadb database backend`

File: separate_instance/docker-compose.override.yml [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+#  #########################################################
+#  ## NGINX ################################################
+#  #########################################################
+#  # nginx:
+#  
+#  #########################################################
+#  ## PHPMYADMIN ###########################################
+#  #########################################################
+#  phpmyadmin:
```
Comment (ulfgebhardt) [inline separate_instance/docker-compose.override.yml:110]: no need to comment this out you can start stuff specifically `docker-compose up mariadb database backend`

File: separate_instance/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      context: ./mariadb
+      target: mariadb_server
+    environment:
+      - MARIADB_ALLOW_EMPTY_PASSWORD=1
+      - MARIADB_USER=root
+      - MYSQL_TCP_PORT=3307
+    networks: 
+      - internal-net
+    ports: 
+      - 3307:3307
```
Comment (ulfgebhardt) [inline separate_instance/docker-compose.yml:76]: if done properly you can override the port with an env variable:

`docker-compose up mariadb database backend --env DATABASE_PORT=3307`

**Review by ulfgebhardt — COMMENTED** [review]
not a fan, I am a friend of properly parameterizing our docker to override the ports used by our services.

File: separate_instance/backend/Dockerfile (hunk 1 above)
Comment (clauspeterhuebner) [reply separate_instance/backend/Dockerfile:1]: ...
ENV PORT="4002"

# Labels
LABEL org.label-schema.name="gradido:backend2"
LABEL org.label-schema.description="Gradido GraphQL Backend 2"
...
LABEL org.label-schema.vendor="Gradido Community 2"
...

File: separate_instance/database/Dockerfile (hunk 2 above)
Comment (clauspeterhuebner) [reply separate_instance/database/Dockerfile:1]: ...
LABEL org.label-schema.name="gradido:database2"
LABEL org.label-schema.description="Gradido Database Migration Service 2"
...
LABEL org.label-schema.vendor="Gradido Community 2"
...

File: separate_instance/database/src/config/index.ts (hunk 3 above)
Comment (clauspeterhuebner) [reply separate_instance/database/src/config/index.ts:1]: ...
  DB_PORT: process.env.DB_PORT ? parseInt(process.env.DB_PORT) : 3307,
...

File: separate_instance/docker-compose.override.yml (hunk 4 above)
Comment (clauspeterhuebner) [reply separate_instance/docker-compose.override.yml:8]: yes you're right, but i had trouble with some modul-dependencies, which cause starting modules automatically, but i just not needed them for my scenario.
Commenting them out was the easiest solution...

File: separate_instance/docker-compose.override.yml (hunk 5 above)
Comment (clauspeterhuebner) [reply separate_instance/docker-compose.override.yml:26]: yes you're right, but i had trouble with some modul-dependencies, which cause starting modules automatically, but i just not needed them for my scenario.
Commenting them out was the easiest solution...

File: separate_instance/docker-compose.override.yml (hunk 6 above)
Comment (clauspeterhuebner) [reply separate_instance/docker-compose.override.yml:110]: yes you're right, but i had trouble with some modul-dependencies, which cause starting modules automatically, but i just not needed them for my scenario.
Commenting them out was the easiest solution...

File: separate_instance/docker-compose.yml (hunk 7 above)
Comment (clauspeterhuebner) [reply separate_instance/docker-compose.yml:76]: yes that's right, but was not the focus of this issue...
even that the previous port definition with 3306 was hardcoded too

Let us define a new issue for: 
- creating scripts to invoke different instances on different ports
- introducing variables in the existing config-, docker- and other-files, which were set outside from the scripts
- ...

all changes of this issue will be part of the root directory "separate_instance" in parallel to all other moduls and can be easiely deleted as soon as we had introduced a dynamic invocation per scripts and config-variables for the concerted modules.

**Comment by clauspeterhuebner** [comment]
close PR: misunderstanding and no requirement in the team to buildup several instances on one host even for development and test... :(

**Pull request closed without merging**